Model calibration

aewallwi opened the pull request:
Add driver function and bash script for model based calibration.

jsdillon commented:
I'm assigning @nkern to this one, since he's the expert in the AbsCal class and will also feel validated by this PR since I argued for taking it out.

nkern commented:
thanks for adding this AEW. a couple of ideas about the structure.

1. phs_logcal still needs a decent starting point, so we should provide the option to perform delay_lincal beforehand (also can you expose dlycal_kwargs={} for the delay lincal step, also should expose similar kwargs for ampcal and phscal)
2. can you expose kwargs for dlycal=True, phscal=True, ampcal=True, such that we can toggle them.
3. should we make the for loop based on a convergence criterion rather than a fixed number of iterations? similar to how redcal step works?

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3694,158 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+    hcm.ref_antenna_name = str(refant)
+
+    for iter in range(iterations):
+        abscal = AbsCal(data=hdd, model=hdm, wgts=wgts)
```
you shouldn't have to reinstantiate the abscal class here. you can simply apply the calibration to the data attached to abscal object

aewallwi replied:
I've rewritten this to update the AbsCal data dictionaries.

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3694,158 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+    hcm = hcm.initialize_from_uvdata(uvdata=hdm, gain_convention='divide', cal_style='sky',
+                                     ref_antenna_name=refant_init, sky_catalog=f'{model_file}',
+                                     metadata_only=False, sky_field=field_str, cal_type='gain')
+    hcm.write_calfits(output_filename + '.tmp')
```
why do we have a temporary file being written and then removed?

aewallwi replied:
The only way I know to initialize a HERACal object is through the filepath to a calfits file -- hence writing the temporary file. If there is a way to initialize HERACal from a uvcal or uvdata/HERAData object in memory this would definitely be the preferable way to go.

nkern replied:
I think perhaps the function you are looking for is `io.to_HERACal` where you can feed a UVCal object rather than a UVCal filename

aewallwi replied:
Yep. Thanks for pointing this out!

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3694,158 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
     return hc
 
 
+def run_model_based_calibration(data_file, model_file, output_filename, auto_file=None,
```
can you also include a pre-calibration input, that multiplies the data_file before calibration starts (and the output calibration should be the delta between the input calibration and the output data)

aewallwi replied:
Will add this.

nkern replied:
on second thought--I think it is more consistent with other objects in redcal if the output calibiration had the input and delta multiplied. maybe its better to have it this way?

aewallwi replied:
Just to clarify, I'm going to multiply the precal gains by the final gain solutions and write those as the output.

aewallwi replied:
I've gone and done this.

aewallwi commented:
> thanks for adding this AEW. a couple of ideas about the structure.
> 
> 1. phs_logcal still needs a decent starting point, so we should provide the option to perform delay_lincal beforehand (also can you expose dlycal_kwargs={} for the delay lincal step, also should expose similar kwargs for ampcal and phscal)
> 2. can you expose kwargs for dlycal=True, phscal=True, ampcal=True, such that we can toggle them.

I've added an option for delay lincal and options to turn off amplitude and phase calibration.

> 3. should we make the for loop based on a convergence criterion rather than a fixed number of iterations? similar to how redcal step works? that being said, the logcal linearization makes the system convex, so is there really a benefit for multiple iterations? perhaps the phase solutions would benefit from this, but I'd have to think about it. do you find a change in the behavior using multiple iterations?

I think the tol idea is much better then iterations. I've changed to this system. I just included iterations as a knob that I thought I might need but never checked if I needed it.

aewallwi commented:
Thanks for all the great input @nkern . I've gone and implemented most of your ideas (using tol instead of iterations and adding options to do amp, phs, and dly) and I think this is ready for another work when you have some time (provided tests pass of course :P)

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+    verbose: bool, optional
+        lots of outputs.
+    """
+    hdd = io.HERAData(data_file)
```
in this case the function `io.to_HERAData` is actually preferred b/c that allows the input `data_file` to be a str or a pre-loaded UVData object, which is nice for sandbox-mode analysis. could you make this change and update the doc string to reflect this?

aewallwi replied:
I'll change this. I also think that ideally `io.HERAData` ought to be able to take UVData objects ;)

aewallwi replied:
I've found that using strings in `to_HERAData` causes some very strange issues (see below). I'm going to stick with `io.HERAData(str)`

```
hera_cal/tests/test_abscal.py Fatal Python error: Illegal instruction

Current thread 0x000000010de52600 (most recent call first):
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/aipy_extracts.py", line 295 in __init__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/miriad.py", line 837 in read_miriad
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 10623 in read_miriad
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 12164 in read
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 11826 in read
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/io.py", line 535 in read
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/abscal.py", line 3748 in run_model_based_calibration
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/tests/test_abscal.py", line 967 in test_run_model_based_calibration
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/python.py", line 1641 in runtest
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 311 in from_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 323 in _main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/config/__init__.py", line 162 in main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/Users/aewallwi/miniconda3/envs/hera/bin/pytest", line 8 in <module>
Illegal instruction: 4
```

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+        # also make sure to only include baselines present in hdd.
+        hdm.select(bls=hdd.bls)
+
+    model, model_flags, model_nsamples = hdm.read()
```
I actually am not sure if this above conditionals are respected when you make a `hdm.read()` call. As i recall from a few years back (this may have changed so @jsdillon please weigh in here), when you do a `hdm.read()` it actually re-loades the filename stored in the object, so all previous antenna / pol selections are actually overwritten.

In order to keep the selections you've done above, you need to use the `_determine_blt_slice` `_determine_pol_slice` and then call `build_datacontainers` directly, which is what the `hdm.read()` call does at the end after re-loading in the file metadata.

nkern replied:
perhaps a test should be constructed to ensure this is working properly too

aewallwi replied:
I've changed the `model, model_flags, model_nsamples = hdm.read()` to `model, model_flags, model_nsamples = hdm.build_datacontainers()`

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+
+    # load in weights if supplied.
+    if auto_file is not None:
+        hda = io.HERAData(auto_file)
```
so `to_HERAData` here

aewallwi replied:
changed.

aewallwi replied:
I've found that using strings in `to_HERAData` causes some very strange issues.

```
hera_cal/tests/test_abscal.py Fatal Python error: Illegal instruction

Current thread 0x000000010de52600 (most recent call first):
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/aipy_extracts.py", line 295 in __init__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/miriad.py", line 837 in read_miriad
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 10623 in read_miriad
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 12164 in read
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pyuvdata/uvdata/uvdata.py", line 11826 in read
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/io.py", line 535 in read
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/abscal.py", line 3748 in run_model_based_calibration
  File "/Users/aewallwi/Dropbox/Aaron/Science/HERA-Team/hera_cal/hera_cal/tests/test_abscal.py", line 967 in test_run_model_based_calibration
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/python.py", line 1641 in runtest
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 311 in from_call
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 323 in _main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/config/__init__.py", line 162 in main
  File "/Users/aewallwi/miniconda3/envs/hera/lib/python3.8/site-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/Users/aewallwi/miniconda3/envs/hera/bin/pytest", line 8 in <module>
Illegal instruction: 4
```

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+            abscal.amp_logcal(verbose=verbose)
+        if not no_phscal:
+            abscal.phs_logcal(verbose=verbose)
+        for k in abscal.ant_eta_gain:
```
this for loop will fail when no_ampcal is True but no_phscal is False b/c ant_eat_gain won't be populated.

you need two for loops, one for eta_gains and one for phi-gain. might as well put them both in the same conditional as the calibration step itself.

```
if not no_ampcal:
    abscal.amp_logcal()
    abscal_gains_iteration = abscal.merge_gains[(abscal_gains_iteration, abscal.ant_eta_gain]
```

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+    hc.ref_antenna_name = str(refant)
+    hcm.ref_antenna_name = str(refant)
+
+    abscal = AbsCal(data=data, model=model, wgts=wgts, freqs=data.freqs)
```
please use a different name for this object, b/c we use `abscal` throughout hera_cal to mean `hc.abscal` module.

I think we may have a precedent for naming class instantiations with uppercase, something like
```
A = AbsCal(...)
```
or 
```
AC = AbsCal(...)
```

aewallwi replied:
Changed to `my_abscal`. There seems to be a mixture of styles for naming instances of classes so I'll keep without caps for now.

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+
+    # multiply gains by precal gains
+    if precalibration_gain_file is not None:
+        hc.gain_array *= uvc_precal.gain_array
```
I think it may be possible that the gain_array antenna ordering in `hc` is different than that of `uvc_precal`. I would suggest eliminating this possible failure mode by either
1. Iterating through hc.ant_array and updating each antenna/pol individually, or 
2. (easier) using `merge_gains([])` to update `abscal_gains` dictionary with `uvc_precal.gain_array` in dictionary form before using `hc.update(gains=abscal_gains)`

aewallwi replied:
I've changed this to use `merge_gains`

nkern reviewed (commented):
couple of more comments, thanks

aewallwi commented on hera_cal/abscal.py:
```diff
@@ -32,6 +32,7 @@
 from scipy import signal, interpolate, spatial
 from scipy.optimize import brute, minimize
 from pyuvdata import UVCal, UVData
+from pyuvdata.utils import uvcalibrate
```
removed.

aewallwi commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+
+    abscal = AbsCal(data=data, model=model, wgts=wgts, freqs=data.freqs)
+
+    delta = 9e99
```
added comment.

aewallwi commented:
Thanks for all the helpful comments @nkern I've added your suggestions and also am dropping the inflate by redundancy options since I don't really need to use them.

nkern commented:
@aewallwi huh thats odd. an easy workaround is something like

def check_str(df):
    if isinstance(df, str):
        return True
    if isinstance(df, (list, tuple, np.ndarray)):
         if isinstance(df[0], str):
            return True

if check_str(auto_file):
    hda = io.HERAData(auto_file)
else:
    hda = io.to_HERAData(auto_file)

nkern commented:
also, is it too hard to keep inflate_by_redundancy an option (perhaps default=False)? I think that's actually a really nice option to have. thanks!

aewallwi commented:
> also, is it too hard to keep inflate_by_redundancy an option (perhaps default=False)? I think that's actually a really nice option to have. thanks!

I've restored the redundancy inflation. I've also fixed some bugs in `to_HERAData` that were preventing metadata from being correctly attached when a `UVData` is provided. This part should be working now. The original errors were because I needed to specify `filetype='uvh5'`, it was trying to read `uvh5` files as `miriad`.

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3693,232 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+                       keep_all_metadata=False)
+        hdm.inflate_by_redundancy()
+        # also make sure to only include baselines present in hdd.
+        hdm.select(bls=hdd.bls)
```
I think you need to follow this select with hdm._determine_blpair_slice and hdm._determine_pol_slice

aewallwi replied:
Added 
```        
hdm._determine_blt_slicing()
hdm._determine_pol_indexing()
```

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3693,232 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+        if not no_ampcal:
+            my_abscal.amp_logcal(verbose=verbose)
+            abscal_gains_iteration = merge_gains([abscal_gains_iteration, my_abscal.ant_eta_gain])
+            # abscal_gains_iteration = {abscal_gains_iteration[k] * my_abscal.ant_eta_gain[k] for k in abscal_gains}
```
rm comment

aewallwi replied:
done

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3693,232 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+        if not no_phscal:
+            my_abscal.phs_logcal(verbose=verbose)
+            abscal_gains_iteration = merge_gains([abscal_gains_iteration, my_abscal.ant_phi_gain])
+            # abscal_gains_iteration = {k: abscal_gains_iteration[k] * my_abscal.ant_phs_gain[k] for k in abscal_gains}
```
same

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3693,232 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+    # multiply gains by precal gains
+    if precalibration_gain_file is not None:
+        abscal_gains = merge_gains([abscal_gains, gains_precal])
+        # abscal_gains = {k: abscal_gains[k] * gains_precal[k] for k in abscal_gains}
```
rm comment

aewallwi replied:
done.

nkern commented on hera_cal/abscal.py:
```diff
@@ -3693,6 +3693,232 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+
+    # find initial starting point with lincal before performing abscal.
+    if dly_lincal:
+        my_abscal.delay_lincal()
```
good to expose kword arguments specifically for delay_lincal, they can have a profound effect on the calibration solution on real data.

Perhaps something like

dlycal_kwargs={} in the signature and then

```
my_abscal.delay_lincal(**dlycal_kwargs)
```

aewallwi replied:
exposed these.

nkern reviewed (approved):
looks good thanks for all the effort on this! should be useful for a variety of use-cases going forward :)
